resources/type-system/inference.md

```diff
@@ -315,45 +320,60 @@ described below with a typing context as computed above.
 The actual returned type of a function literal with a block body is computed as
 follows.  Let `T` be `Never` if every control path through the block exits the
 block without reaching the end of the block, as computed by the **definite
-completion** analysis specified elsewhere.  Let `T` be `Null` if any control
+completion** analysis specified elsewhere, or if the function is a generator
+function.
+Let `T` be `Null` if the function is a non-generator function and any control
 path reaches the end of the block without exiting the block, as computed by the
 **definite completion** analysis specified elsewhere.  Let `K` be the typing
 context for the function body as computed above from the imposed return type
 schema.
-  - For each `return e;` statement in the block, let `S` be the inferred type of
-    `e`, using the local type inference algorithm described below with typing
-    context `K`, and update `T` to be `UP(flatten(S), T)` if the enclosing
-    function is `async`, or `UP(S, T)` otherwise.
-  - For each `return;` statement in the block, update `T` to be `UP(Null, T)`.
+  - If the enclosing function is a non-`async` non-generator function,
+    for each `return e;` statement in the block, let `S` be the inferred type
+    of `e`, using the local type inference algorithm described below with typing
+    context `K`, and update `T` to be `UP(S, T)`.
+  - If the enclosing function is marekd `async`, for each `return e;` statement
+    in the block, let `S` be the inferred type of `e`, using the local type
+    inference algorithm described below with typing context `FutureOr<K>`,
+    and update `T` to be `UP(flatten(S), T)`.
+  - If the enclosing function is a non-generator function, for each `return;`
+    statement in the block, update `T` to be `UP(Null, T)`.
   - For each `yield e;` statement in the block, let `S` be the inferred type of
     `e`, using the local type inference algorithm described below with typing
     context `K`, and update `T` to be `UP(S, T)`.
   - If the enclosing function is marked `sync*`, then for each `yield* e;`
     statement in the block, let `S` be the inferred type of `e`, using the
     local type inference algorithm described below with a typing context of
-    `Iterable<K>`; let `E` be the type such that `Iterable<E>` is a
-    super-interface of `S`; and update `T` to be `UP(E, T)`.
+    `Iterable<K>`. If there exists a type `E` such that `Iterable<E>` is a
+    super-interface of `S`, update `T` to be `UP(E, T)`. Otherwise update
+    `T` to be `UP(S, T)`.
+    _It is a compile-time error if *S* is not a assignable to
```

Comment:
This seems reasonable, the previous behavior was underspecified.  Note that the analyzer and the CFE currently disagree here, so there is an implementation change implied.  Test case to reveal the difference:
```dart
void main() {
  var f = () sync* {
    yield* (throw "hello");
  };
  int x = f();
  var g = () sync* {
    yield* (3 as dynamic);
  };
  int y = g();
}
```